contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -6391,6 +6579,690 @@ bbf_ExecDropStmt(DropStmt *stmt)
+ *	For a rename operation on a babelfish partitioned table, it renames all partition
+ *	of the table to the new name using hash based on the new name, so that
+ *	name of partitions of any partitioned table doesn't conflict.
+ */
+static void
+rename_table_update_bbf_partitions_name(RenameStmt *stmt)
+{
+	char		*physical_schema_name = stmt->relation->schemaname;
+	char		*logical_schema_name;
+	char		*table_name = stmt->relation->relname;
+	List		*partition_names = NIL;
+	int16		dbid;
+	char		*unique_hash;
+	PlannedStmt	*wrapper;
+	RenameStmt	*rename_partition_stmt = NULL;
+	List		*parsetree;
+	Relation	relation;
+	SysScanDesc	scan;
+	ScanKeyData	key;
+	Oid		parentrelid;
+	Oid		inhrelid;
+	HeapTuple	tuple;
+	const char	*old_dialect;
+	StringInfoData	query;
+	char		*partition_name;
+	char		*new_partition_name;
+
+	logical_schema_name = (char *) get_logical_schema_name(physical_schema_name, true);
+
+	if (!logical_schema_name) /* not a tsql schema */
+		return;
+	
+	dbid = get_dbid_from_physical_schema_name(physical_schema_name, false);
+	
+	if (!is_bbf_partitioned_table(dbid, logical_schema_name, table_name))
+	{
+		pfree(logical_schema_name);
+		return;
+	}
+
+	parentrelid = get_relname_relid(stmt->newname, get_namespace_oid(physical_schema_name, false));
+
+	relation = table_open(InheritsRelationId, AccessShareLock);
+
+	ScanKeyInit(&key,
+			Anum_pg_inherits_inhparent,
+			BTEqualStrategyNumber, F_OIDEQ,
+			ObjectIdGetDatum(parentrelid));
+
+	scan = systable_beginscan(relation, InheritsParentIndexId, true,
+					NULL, 1, &key);
+
+	while ((tuple = systable_getnext(scan)) != NULL)
+	{
+		inhrelid = ((Form_pg_inherits) GETSTRUCT(tuple))->inhrelid;
+		partition_names = lappend(partition_names, get_rel_name(inhrelid));
+	}
+	systable_endscan(scan);
+	table_close(relation, AccessShareLock);
+
+	/*
+	 * We prepare the following query to rename the partitions of partitioned table.
+	 * This will be executed using ProcessUtility().
+	 */
+	initStringInfo(&query);
+	appendStringInfo(&query, "ALTER TABLE %s.dummy RENAME TO dummy;", physical_schema_name);
+
+	old_dialect = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+
+	/* We need to change the dialect to postgres to parse the RENAME statement. */
+	PG_TRY();
+	{
+		set_config_option("babelfishpg_tsql.sql_dialect", "postgres", GUC_CONTEXT_CONFIG,
+					PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+
+		parsetree = raw_parser(query.data, RAW_PARSE_DEFAULT);
+		rename_partition_stmt = (RenameStmt *) parsetree_nth_stmt(parsetree, 0);
+	}
+	PG_FINALLY();
+	{
+		/* Reset dialect back to original value. */
+		set_config_option("babelfishpg_tsql.sql_dialect", old_dialect, GUC_CONTEXT_CONFIG,
+					PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	}
+	PG_END_TRY();
+
+	/* Need to make a wrapper PlannedStmt. */
+	wrapper = makeNode(PlannedStmt);
+	wrapper->commandType = CMD_UTILITY;
+	wrapper->canSetTag = false;
+	wrapper->utilityStmt = (Node *) rename_partition_stmt;
```

Comment:
It is possible that rename_partition_stmt might be NULL here, we should detect such cases

Reply:
No, it is not possible that `rename_partition_stmt` to be NULL. The `parsetree_nth_stmt` will lead assertion error in case of any such failure.